### DISCUSSION TITLE
[@types/aws-lambda] `Types of parameters 'event' and 'event' are incompatible.`

### DISCUSSION THREAD

**Discussion by garysassano** [DefinitelyTyped/DefinitelyTyped#67933] [Issues with a @types package] [answered]
![image](https://github.com/aws/aws-sdk-js-v3/assets/10464497/4905a6ef-6182-4cab-a369-93f223f74660)

Full error:
```
Type '(event: S3Event) => Promise<APIGatewayProxyResult>' is not assignable to type 'APIGatewayProxyHandler'.
  Types of parameters 'event' and 'event' are incompatible.
    Property 'Records' is missing in type 'APIGatewayProxyEventBase<APIGatewayEventDefaultAuthorizerContext>' but required in type 'S3Event'.
```

[a comment by an automated account is not shown]

**Comment by simonbuchan** [comment]
Why would you expect an S3 Event to be usable in an API gateway handler?

**Reply by garysassano** [reply]
I was just pointing out how the error message is ambiguous.

**Reply by garysassano** [reply]
Btw, why there's a `S3Handler`, `S3Event`, but only `S3BatchResult`, without any `S3Result`? And what is `S3BatchResultResult`?

**Reply by garysassano** [reply]
Is it more appropriate to use `S3NotificationEvent` instead of `S3Event` when an object is being added to an S3 bucket? Furthermore, would `S3ObjectCreatedNotificationEvent` be an even more suitable choice for such a scenario?

**Reply by simonbuchan** [reply, marked as answer]
The very short tldr is just use the Handler types and let Typescript infer the event and result types: you can just use `export const handler: FooHandler = async (event) => ...` and it all works out. The handler you use must match the AWS trigger that will call the lambda, the comments on the handler or event types should describe which, and provide a link to the service documentation describing the trigger event and result payloads.

Not every trigger has a meaningful Result type, but otherwise the intent is that `type FooHandler = Handler<FooEvent, FooResult>`, so if there's no Result type, that (should be) because there's no result to return, the handler returns void.

Can't do much about the error, that's Typescript. It's correct, though somewhat confusing if you don't know what it's doing: in order for you to assign an expression of some type `(A1) => R1` to a variable (or argument, etc) with type `(A2) => R2`, Typescript checks that R1 is assignable to R2 (because the actual function will return an R1, but the caller will expect an R2), but that the argument A2 is assignable to A1, because the caller can provide any A2, but the actual function only expects an A1. (This is called covariance and contravariance respectively, if you want to search for more information about this)

In this case Typescript could more clearly say that the types S3Event and APIGatewayProxyEvent are unrelated rather than picking a random property to complain about, but good errors are hard.

I expect that S3BatchResultResult is the type of a result property of S3BatchResult.
